### DISCUSSION TITLE
Add ability to save address after checkout

### DISCUSSION THREAD

**Discussion by rynpsc** [craftcms/commerce#2843] [Ideas] [answered]
With the address updates in Craft 4 addresses are no longer saved during checkout. It would be nice if during checkout when a logged in user is adding a new shipping or billing address they could chose to save that address for future use in the same way you can save a payment source during checkout via `savePaymentSource`, e.g. `saveShippingAddress` and `saveBillingAddress`.

**Comment by nfourtythree** [comment]
Hi @rynpsc 

Thank you for the suggestion. This is something we will take into consideration.

In the meantime, you could add something similar to the following code to a custom module to get your desired behaviour.

In the example below I have made the assumption that you would create a custom field on the order that will be storing whether the customer would like to save their addresses at the end of the checkout process.

```php
Event::on(Order::class, Order::EVENT_AFTER_COMPLETE_ORDER, static function(Event $event) {
    /** @var Order $order */
    $order = $event->sender;

    // Custom logic here to determine if to save the addresses or not
    if ($order->mySaveAddressesCustomField) {
        if ($billingAddress = $order->getBillingAddress()) {
            Craft::$app->getElements()->duplicateElement($billingAddress, ['ownerId' => $order->getCustomerId()]);
        }

        if ($shippingAddress = $order->getShippingAddress()) {
            Craft::$app->getElements()->duplicateElement($shippingAddress, ['ownerId' => $order->getCustomerId()]);
        }
    }
});
```

Hope this helps, thanks!

**Reply by tmjoen** [reply]
This was almost exactly what I needed, except I can't seem to access my custom order field. I am a little unsure if I'm sending it correctly through my frontend.

I added a lightswitch field `saveShippingAddress` to my order model, then in my frontend have this in a form

```
{{ input('checkbox', 'fields[saveShippingAddress]', 1, {
  id: 'saveShippingAddress',
}) }}
```

and check it in the event like this:

```php
    Event::on(Order::class, Order::EVENT_AFTER_COMPLETE_ORDER, static function(Event $event) {
            /** @var Order $order */
            $order = $event->sender;
            
            if ($order->saveShippingAddress) {
                if ($shippingAddress = $order->getShippingAddress()) {
                    Craft::$app->getElements()->duplicateElement($shippingAddress, ['ownerId' => $order->getCustomerId()]);
                }
            }
        });
```

Thanks for any insight :)

**Reply by nfourtythree** [reply]
Hi @tmjoen 

After you save the `saveShippingAddress` value during the checkout have you checked on the control panel that the light switch is set to "on"?

Thanks

**Comment by bleepsandblops** [comment]
Just to add to this, the sample templates are quite confusing, as they include fields like `Make this my default shipping address` but without adding this little bit of code the address doesn't save.

**Reply by MattWilcox** [reply]
OMG this. What is the point of these options if they literally do nothing?

**Reply by pixelmachine** [reply]
I think this is now fixed in 4.3 @MattWilcox 

https://craftcms.com/docs/commerce/4.x/addresses.html#save-addresses-when-completing-an-order

I guess you need to save the address and send `makePrimaryBillingAddress` at the same time? Docs aren't clear.

Correct @nfourtythree ?

Looks like example templates weren't updated.

**Comment by emarthinsen** [comment]
We use this feature to save an address for future use pretty heavily in our 3.x Commerce installation. It would be nice if this feature came back.

**Comment by lukeholder** [comment, marked as answer]
This is possible with the `saveBillingAddressOnOrderComplete` and `saveShippingAddressOnOrderComplete` cart update params.

There is also an example of this in the options page of the checkout flow in the example templates.
